src/Integration/LocalServices/UnboundProjectFinder.cs

```diff
@@ -76,31 +84,80 @@ public IEnumerable<Project> GetUnboundProjects()
 
         private Project[] GetUnboundProjects(BindingConfiguration binding)
         {
+            threadHandling.ThrowIfOnUIThread();
+
             Debug.Assert(binding.Mode.IsInAConnectedMode());
             Debug.Assert(binding.Project != null);
 
-            var filteredSolutionProjects = projectSystem.GetFilteredSolutionProjects().ToArray();
-            logger.LogDebug($"[Binding check] Number of bindable projects: {filteredSolutionProjects.Length}");
+            // Using threadHandling.Run(...) here to avoid having to convert all of the upstream
+            // // calling methods to be async.
+            var projects = threadHandling.Run(() => GetUnboundProjectsAsync(binding));
+
+            threadHandling.ThrowIfOnUIThread();
+            return projects;
+        }
 
-            var result = filteredSolutionProjects
```

Comment:
The logic is the same as in the previous `Where` clause. However, each inner iteration involves a thread switch to the UI thread.